### PR TITLE
Simplify JsonStream, avoid unboxing errors, propagate exceptions more sanely, add test

### PR DESCRIPTION
**Pull request by loopj**

### CODE REVIEW

File: src/androidTest/java/com/bugsnag/android/BugsnagTestCase.java [hunk 1]
```diff
@@ -13,7 +13,9 @@ protected String streamableToString(JsonStream.Streamable streamable) {
         StringWriter writer = new StringWriter();
         JsonStream jsonStream = new JsonStream(writer);
 
-        streamable.toStream(jsonStream);
+        try {
+            streamable.toStream(jsonStream);
+        } catch(java.io.IOException e) {}
```
Comment (JakeWharton) [inline src/androidTest/java/com/bugsnag/android/BugsnagTestCase.java]: nit: import?

File: src/main/java/com/bugsnag/android/AppData.java [hunk 2]
```diff
@@ -32,7 +32,7 @@
         guessedReleaseStage = guessReleaseStage();
     }
 
-    public void toStream(JsonStream writer) {
+    public void toStream(JsonStream writer) throws java.io.IOException {
```
Comment (JakeWharton) [inline src/main/java/com/bugsnag/android/AppData.java]: nit: import?

Comment (JakeWharton) [inline src/main/java/com/bugsnag/android/AppData.java]: there's a lot of these...

File: src/main/java/com/bugsnag/android/ErrorStore.java [hunk 3]
```diff
@@ -76,7 +76,9 @@ void write(Error error) {
         String filename = String.format("%s%d.json", path, System.currentTimeMillis());
         try {
             Writer out = new FileWriter(filename);
```
Comment (JakeWharton) [inline src/main/java/com/bugsnag/android/ErrorStore.java]: Unrelated, but `Writer out` should be outside the try/catch and a null check + close in `finally`

**Comment by loopj** [comment]
Certainly reasonable to import all those.

Also I'll actually need to re-add some null-protection for `Boolean` etc since the underlying `JsonWriter` only accepts the primitive versions.

File: src/main/java/com/bugsnag/android/JsonStream.java [hunk 4]
```diff
@@ -5,135 +5,41 @@
...
-            }
+    // Add support for File values
+    public void value(File file) throws IOException {
+        super.flush();
 
-            out.flush();
-        } catch (IOException e) {
-            e.printStackTrace();
+        // Buffer the file contents onto the stream
+        FileReader input = new FileReader(file);
```
Comment (JakeWharton) [inline src/main/java/com/bugsnag/android/JsonStream.java]: Needs try/finally for close.

**Comment by loopj** [comment]
Thanks for the comments @JakeWharton !

**Comment by loopj** [comment]
Pushed my changes from last night, ready to merge. @snmaynard @ConradIrwin want to review?

**Pull request merged**